mlrun/projects/project.py

```diff
@@ -2336,6 +2346,121 @@ def build_function(
+        requirements: typing.Union[str, typing.List[str]] = None,
+        mlrun_version_specifier=None,
+        builder_env: dict = None,
+        overwrite_build_params: bool = False,
+    ) -> typing.Union[BuildStatus, kfp.dsl.ContainerOp]:
+        """Builder docker image for the project, based on the project's build config. Parameters allow to override
+        the build config.
+
+        :param image: target image name/path. If not specified the project's existing `default_image` name will be
+                        used. If not set, the `mlconf.default_project_image_name` value will be used
+        :param set_as_default: set `image` to be the project's default image (default False)
+        :param with_mlrun:      add the current mlrun package to the container build
+        :param skip_deployed:   skip the build if we already have the image specified built
+        :param base_image:      base image name/path (commands and source code will be added to it)
+        :param commands:        list of docker build (RUN) commands e.g. ['pip install pandas']
+        :param secret_name:     k8s secret for accessing the docker registry
+        :param requirements:    list of python packages or pip requirements file path, defaults to None
+        :param mlrun_version_specifier:  which mlrun package version to include (if not current)
+        :param builder_env:     Kaniko builder pod env vars dict (for config/credentials)
+                                e.g. builder_env={"GIT_TOKEN": token}, does not work yet in KFP
+        :param overwrite_build_params:  overwrite existing build configuration (default False)
+
+           * False: the new params are merged with the existing (currently merge is applied to requirements and
+             commands)
+           * True: the existing params are replaced by the new ones
+        """
+
+        self.build_config(
+            image=image,
+            set_as_default=set_as_default,
+            base_image=base_image,
+            commands=commands,
+            secret_name=secret_name,
+            with_mlrun=with_mlrun,
+            requirements=requirements,
+            overwrite_build_params=overwrite_build_params,
+        )
+
+        function = mlrun.new_function("dummy--image--build", kind="job")
+
+        build = self.spec.build
+        result = self.build_function(
+            function=function,
+            with_mlrun=build.with_mlrun,
+            image=build.image,
+            base_image=build.base_image,
+            commands=build.commands,
+            secret_name=build.secret,
+            requirements=build.requirements,
+            skip_deployed=skip_deployed,
+            overwrite_build_params=overwrite_build_params,
+            mlrun_version_specifier=mlrun_version_specifier,
+            builder_env=builder_env,
+        )
+
+        mlrun.db.get_run_db(secrets=self._secrets).delete_function(
+            name=function.metadata.name
+        )
```

Comment:
not sure I understand the entire reason why you create a function and then delete it. do you mind documenting it? this behavior / flow needs to be well explained

Reply:
also, what happens when new function / delete function fails? do we want to blow the entire flow? leave a warning and still return the results? etc perhaps the delete function can happen in the background of api?

Reply:
The building process is all centered around a runtime for which the image is built. This is true for the SDK side as well as REST API + server side processing. Generally speaking it should be enough to pass the `ImageBuilder` configuration + perhaps some additional details to this process, but for the purpose of this PR this seems like an overkill (not to mention having to sort out BC issues that may arise).
For now the build process of the project object is just a wrapper to automate this "make a runtime, build image, delete runtime" process (which the user could have done previously, of course). The big advantage of this PR is actually the addition of the build config to the project, such that it is exported and is persisted with the project.

Reply:
I see your point around the delete-function, as it is not strictly needed for the build flow, I'll make it optional (if it fails there will be leftovers, but the build will be successful). I don't see a lot of value in making it run in the background since it's just deleting the function object from DB, and it adds a non-significant amount of time to the build process.